File: c/forward_difference.c
#include <stdio.h>
#include <stdbool.h>

void forward(int items[], size_t n, int times) {
    if (times == 0) {
        return;
    }

    for (int i = 0; i < n; i++) {
        printf("%d", items[i]);
        if (i != n - 1) {
            printf(", ");
        }
    }

    int new_lis[n - 1];

    for (int i = 0; i < n - 1; i++) {
        new_lis[i] = items[i+1] - items[i];
    }

    printf("\n");
    return forward(new_lis, n - 1, times - 1);
}

int main() {
    int vars[] = {90, 47, 58, 29, 22, 32, 55, 5, 55, 73};
    size_t size = sizeof(vars)/sizeof(int);

    forward(vars, size, 10);
    return 0;
}

File: python/forward_difference.py
# https://rosettacode.org/wiki/Forward_difference

def forward(items: list[int], *, times: int = 1, display_all: bool = False) -> list[int]:
    if times == 0:
        return items

    new = [0]*(len(items)-1)
    for i, item in enumerate(items[:-1]):
        new[i] = items[i+1] - item

    if display_all:
        print(new)
    return forward(new, times=times-1, display_all=display_all)

def main():
    forward([90, 47, 58, 29, 22, 32, 55, 5, 55, 73], times=9, display_all=True)

if __name__ == '__main__':
    main()
